### ISSUE TITLE
Adding a shadow to the image

### ISSUE THREAD

**Issue by niravasher** [publiclab/image-sequencer#806]
### Please describe the problem (or idea)

> What did you expect to see that you didn't?
Add a shadow to the image

### Please show us where to look
http://sequencer.publiclab.org/examples/#steps=

### What's your PublicLab.org username?
niravasher

### Browser, version, and operating system
windows 10 with ubuntu subsystem
****

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site.

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders
@jywarren

**Comment by niravasher** [comment]
Hi, please tell how to proceed with this? @jywarren @VibhorCodecianGupta

**Comment by jywarren** [comment]
Can you copy in my last comment about meta modules here? There should be
directions in the readme now for meta module creating, and I believe the
colorbar module is an example. Basically it's a module made of other
modules, and if you try out the set I mentioned, I think you can make an
initial attempt at this using a meta module!

**Comment by niravasher** [comment]
Hello, do I need to load Image inside this module? @jywarren @HarshKhandeparkar ?How to decide in which module, image has to be loaded?

**Comment by niravasher** [comment]
Also I made a new module, kept its inBrowser flag and ui flag to true, made index.js, info.json and Module.js. Required this module in the src/Modules.js file. Is there anything else that I need to do to test that module in my local, because I can't see the shadow module in "more modules" in my browser.

**Comment by HarshKhandeparkar** [comment]
Nirav can you please open a PR? We can discuss and debug it there.

**Comment by niravasher** [comment]
Umm.... I am sorry, I didn't get you @HarshKhandeparkar , I can't see shadow module on my local, why should I open a PR?

**Comment by HarshKhandeparkar** [comment]
Niravasher I asked you to open a PR so that I can have a look at the code and debug any errors.

**Comment by niravasher** [comment]
Opened a PR , https://github.com/publiclab/image-sequencer/pull/816

**Comment by niravasher** [comment]
Hey, I was working since yesterday. This is what I have made till now
![Screenshot (47)](https://user-images.githubusercontent.com/32176296/54780621-3d34d000-4c40-11e9-8376-5513258fa842.png)
Now I am trying to add shadow on the top portion of the image. Is this relevant? @jywarren @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Ya good going. Improvements can be made(if reqd) once the scaffolding is done.

**Comment by niravasher** [comment]
What all options should we give to the user in this module?
I mean shadow in the top , left , right , bottom or two sided shadow, three sided shadow, all sided shadow?
As in what should I add in info.json file @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Maybe the same as CSS shadows use?

**Comment by niravasher** [comment]
CSS shadows also give blur effect. That is a bit difficult but will try to do that (maybe i have to change the alpha value of the gradient). So there will be an option for three inputs separated by a space @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
You can implement blur in a follow up pr.

**Comment by niravasher** [comment]
Ok, roger that. Also how do we add color to the shadow? @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Is it of any use?

**Comment by niravasher** [comment]
CSS has the option, maybe its not of any use. So it will be just a two input space separated option

**Comment by HarshKhandeparkar** [comment]
You can add some color let's say #RRGGBB, then you can decrease each of the channels proportionally.

**Comment by HarshKhandeparkar** [comment]
It will be a string with space separted decimal values for each channel.

**Comment by niravasher** [comment]
I don't think a normal person who is not a developer and just wants a shadow would know what is a hexcode

**Comment by HarshKhandeparkar** [comment]
Ya you can use hex but other modules use decimal. Also we use decimal values for channels in PixelManipulation so parsing hex into decimal will be difficult.

**Comment by niravasher** [comment]
Ok for now I am keeping only two inputs, will change anything in the follow up PR. Is that Ok?

**Comment by HarshKhandeparkar** [comment]
:+1:

**Comment by HarshKhandeparkar** [comment]
I think when x+25 is not inside the image. I.e. greater then the width of the image and lies outside of it, it throws an error and doesn't work.

**Comment by niravasher** [comment]
Doesn't throw any erro in the console @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Any screenshots of what it does? Secondly have you built the dist and cleared the cache?

**Comment by niravasher** [comment]
This is not doing anything. Yes I built the dist and cleared the cache @HarshKhandeparkar

**Comment by jywarren** [comment]
Hi, all - couple notes here - we have an improvement issue for "color
selection input" so that will simplify color selection (but we can do it
later). Second, are you increasing the canvas size? Because I
believe @tech4gt will be wanting such a step too for unrelated reasons, and
perhaps we should share code on that!

This is looking good -- perhaps for the time being you could test it with
an image that has a transparent background, like this one:
http://www.clker.com/cliparts/5/x/Q/n/Q/p/alpha-md.png and then not worry
about canvas size for now. We could do canvas size in a separate module,
maybe?

**Comment by niravasher** [comment]
I am going for resize now. Will do that later. Right now I am moving the image to the right so that it does not get cropped when the gradient is applied

**Comment by niravasher** [comment]
I made a new js file inside the folder _nomodule. I can now translate the image any number of pixels in any direction and it won't be cut.
![Screenshot (50)](https://user-images.githubusercontent.com/32176296/54863738-56d63480-4d72-11e9-8b11-a4592c6e51f9.png)

The next step is, I am applying the gradient and opening a PR today itself.
After that will work on resizing the canvas @jywarren @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
:+1: what is that file called? Can the same func be added to PixelManipulation? Like a function translatePix(x, y) just like others like getNeighboutPixel, extraManipulation, changePixel? Or alternatively, you can require that file in PixelManipulation and let PixelManipulation return the func. If you have already done this then it's ok.

**Comment by niravasher** [comment]
I am trying to add it to pixelManipulation

**Comment by niravasher** [comment]
Yay! Pixel manipulation is modified, we have a shadow 😄 
![Screenshot (51)](https://user-images.githubusercontent.com/32176296/54863962-70c54680-4d75-11e9-8b53-3b30516367c1.png)
@HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Wow! Can the shadow be oy added in the top and left sides or anywhere? (I mean can the function oy translate the img in +ve directions or also -ve)?

**Comment by niravasher** [comment]
I can translate the image in any direction and also can add the gradient in any direction. Shadow can be one sided or two sided but can't be three sided or four sided. @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Ok great!

**Comment by niravasher** [comment]
I am stuck here. I have to call changePosition function every time when I press apply button, this is only possible when I keep the function inside draw function. So I removed it from pixelManipulation and took it in Module.js. Here I declared an array called pixelArray. All the channels of all pixels are stored in this array. Now I do pixels.set(x-firstValue, y-secondValue, channel, pixelArray[i]) . This shifts the pixels by firstValue and secondValue. But this is not working, opening up a PR, please help me debug there.
@HarshKhandeparkar @jywarren

**Comment by HarshKhandeparkar** [comment]
PixelManipulation does run each time you press the apply button. You just put PixelManipulation inside draw.

**Comment by jywarren** [comment]
Hi @niravasher can you share a PR? are you using the blur module to generate the shadow? 

This is looking good, but I was thinking we really could make a metamodule if we debug the other modules to follow a path like this:

https://sequencer.publiclab.org/examples/#steps=threshold{},blur{},overlay{offset:-4}

See, here, we:

1. threshold the image to make a black and white version
2. then we blur the black and white version to make the shadow
3. then we overlay the original image over the blurry shadow

But a few missing features or bugs are stopping us:

1. blur does not have a radius #944
2. overlay seems not to be working
3. blurring an image with alpha channel seems not to work; we need to blur the alpha channel too, maybe?

See how this might be made to work, though? 

I think also that making a `canvas-size` module which can increase the canvas size, would be a good idea to make this easier. 

Thanks for your hard work, @niravasher !
